src/main/java/winstone/Launcher.java

```diff
@@ -229,9 +231,19 @@ private void writePortToFileIfNeeded() throws IOException {
                     Path portFile = Paths.get(portFileName);
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
-                    try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                    Path tmpPath = Files.createTempFile(portDir, "temp", "");
+                    try (BufferedWriter writer = Files.newBufferedWriter(tmpPath, StandardCharsets.UTF_8)) {
                         writer.write(Integer.toString(port));
                     }
+
+                    try {
+                        Files.move(tmpPath, portFile, StandardCopyOption.ATOMIC_MOVE);
+                    } catch (AtomicMoveNotSupportedException e) {
```

Comment:
You may also need to catch `IOException` in case the AtomicMOve operation s*is* supported, yet it is not a valid OS/filesystem combination that can overwrite an existing file.

See https://github.com/jenkinsci/jenkins/blob/0907d01d0a9e567b81dfca88ad854b5ef6dc9577/core/src/main/java/hudson/util/AtomicFileWriter.java#L193-L224 for some battle tested code.

If the file is not expected to exist this can be simplified somewhat but per the comment above from Basil and the code below wrt `REPLACE_EXISTING` it  am not sure if we expect it or not.

Reply:
YAGNI I think. While I don't doubt that there were some truly bizarre edge cases that led to the battle-tested code in core's `AtomicFileWriter`, I think that is overkill in this case and that it would be preferable to keep it simple and not deal with too many pathological cases other than the usual `AtomicMoveNotSupportedException`.